docs/architecture/architecture.md

```diff
@@ -0,0 +1,378 @@
+to the container process.
+In the Clear Containers context, this means `cc-runtime` needs a way to send a signal
+to the container process within the virtual machine. As `cc-shim` is responsible for
+forwarding signals to its associated running containers, `cc-runtime` naturally
+calls `kill` on the `cc-shim` PID.
+
+However, `cc-shim` is not able to trap `SIGKILL` and `SIGSTOP` and thus `cc-runtime`
+needs to follow a different code path for those 2 signals.
+Instead of `kill`'ing the `cc-shim` PID, it will go through the following steps:
+
+1. `cc-runtime` connects to `cc-proxy` and sends it the `attach` command to let it know
+on which pod the container it is trying to `kill` is running.
+2. `cc-runtime` sends an agent `KILLCONTAINER` command to `kill` the container running
+on the guest. The command is sent to `cc-proxy` who forwards it to the right agent instance
+running in the appropriate guest.
+
+##### `delete`
+
+`docker delete` is about deleting all resources held by a stopped/killed container. Running
+containers can not be `delete`d unless the OCI runtime is explictly being asked to. In that
+case it will first `kill` the container and only then `delete` it.
+
+The resources held by a Clear Container are quite different from the ones held by a host
+namespace container e.g. run by `runc`. `cc-runtime` needs mostly to delete the pod
+holding the stopped container on the virtual machine, shut the hypervisor down and finally
+delete all related proxy resources:
+
+1. `cc-runtime` connects to `cc-proxy` and sends it the `attach` command to let it know
+on which pod the container it is trying to to `delete` is running.
+2. `cc-runtime` sends an agent `DESTROYPOD` command to `destroy` the pod holding the
+container running on the guest. The command is sent to `cc-proxy` who forwards it to the right
+agent instance running in the appropriate guest.
+3. After deleting the last running pod, the `agent` will gracefully shut the virtual machine
+down.
+4. `cc-runtime` sends the `BYE` command to `cc-proxy`, to let it know that a given virtual
+machine is shut down. `cc-proxy` will then clean all its internal resources associated with this
+VM.
+
+#### Proxy
+
+`cc-proxy` is a daemon offering access to the VM [`agent`](https://github.com/clearcontainers/agent)
+to multiple `cc-shim` and `cc-runtime` clients.
+Only a single instance of `cc-proxy` per host is necessary as it can be used for several different VMs.
+Its main role is to:
+- Arbitrate access to the `agent` control channel between all the `cc-runtime` instances and the `cc-shim` ones.
+- Route the I/O streams between the various `cc-shim` instances and the `agent`.
+
+`cc-proxy` provides 2 client interfaces:
+
+- A UNIX, named socket for all `cc-runtime` instances on the host to send commands to `cc-proxy`.
+- One socket pair per `cc-shim` instance, to send stdin and receive stdout and stderr I/O streams. See the
+[cc-shim section](#shim)
```

Comment:
That has changed as well and is now simplified. There is one single socket (UNIX or TCP) for all runtimes and all shims.

Reply:
@jcvenegas -- can you take a look and update?

Reply:
@egernst  I just check it is updated describing  the interaction between proxy and shim in the new protocol